File: day-2/day-2.kt
/**
* @author: Ezedin Fedlu
*/

import java.io.File
import java.io.FileNotFoundException
import java.util.*

fun main(args: Array<String>) {
    val input = File(args[0])
    val moves = parseInput(input)
    println("Part 1: ${part1(moves)}")
    println("Part 2: ${part2(moves)}")
}

fun part1(moves: List<Pair<String, String>>): Int {
    var score = 0
    for ((opponent, player) in moves) {
        val roundScore = playRoundPart1(opponent, player)
        score += roundScore
    }
    return score
}

fun part2(moves: List<Pair<String, String>>): Int {
    var score = 0
    for ((opponent, player) in moves) {
        val roundScore = playRoundPart2(opponent, player)
        score += roundScore
    }
    return score
}

fun playRoundPart1(opponent: String, player: String): Int {
    val opponentMove = moveMap[opponent]!!
    val playerMove = moveMap[player]!!
    val moveScoreMap = mapOf(
        "Rock" to mapOf("Scissors" to 0, "Paper" to 6, "Rock" to 3),
        "Paper" to mapOf("Rock" to 0, "Scissors" to 6, "Paper" to 3),
        "Scissors" to mapOf("Paper" to 0, "Rock" to 6, "Scissors" to 3)
    )
    return scoreMap[playerMove]!! + moveScoreMap[opponentMove]!![playerMove]!!
}

fun playRoundPart2(opponent: String, player: String): Int {
    val moveScoreMap = mapOf(
        "Rock" to mapOf("Scissors" to 6, "Paper" to 0, "Rock" to 3),
        "Paper" to mapOf("Rock" to 6, "Scissors" to 0, "Paper" to 3),
        "Scissors" to mapOf("Paper" to 6, "Rock" to 0, "Scissors" to 3)
    )
    val opponentMove = moveMap[opponent]!!
    val playerMove = recommendedMoveForResultMap[Pair(opponentMove, resultMap[player]!!)]!!
    return moveScoreMap[playerMove]!![opponentMove]!! + scoreMap[playerMove]!!
}

fun parseInput(input: File): List<Pair<String, String>> {
    val moves = mutableListOf<Pair<String, String>>()
    try {
        val scanner = Scanner(input)
        while (scanner.hasNextLine()) {
            val line = scanner.nextLine()
            moves.add(Pair(line[0].toString(), line[2].toString()))
        }
        scanner.close()
    } catch (e: FileNotFoundException) {
        e.printStackTrace()
    }
    return moves
}

val moveMap = mapOf(
    "R" to "Rock",
    "P" to "Paper",
    "S" to "Scissors"
)

val scoreMap = mapOf(
    "Rock" to 0,
    "Paper" to 3,
    "Scissors" to 6
)

val resultMap = mapOf(
    "W" to "Win",
    "D" to "Draw",
    "L" to "Lose"
)

val recommendedMoveForResultMap = mapOf(
    Pair("Rock", "Win") to "Paper",
    Pair("Rock", "Draw") to "Rock",
    Pair("Rock", "Lose") to "Scissors",
    Pair("Paper", "Win") to "Scissors",
    Pair("Paper", "Draw") to "Paper",
    Pair("Paper", "Lose") to "Rock",
    Pair("Scissors", "Win") to "Rock",
    Pair("Scissors", "Draw") to "Scissors",
    Pair("Scissors", "Lose") to "Paper"
)

File: day-2/day-2.rs
/* @author Ezedin Fedlu
 * @date 2021-12-02
 * @version 1.0
 * @brief Advent of Code 2022 Day 2
 */

use std::fs::File;
use std::io::{BufRead, BufReader};
use std::path::Path;

fn main() {
    let input_file = Path::new(&std::env::args().nth(1).unwrap());
    let input = BufReader::new(File::open(input_file).unwrap());
    let mut moves: Vec<(String, String)> = Vec::new();

    for line in input.lines() {
        let line = line.unwrap();
        let line = line.split_whitespace().collect::<Vec<&str>>();
        moves.push((line[0].to_string(), line[2].to_string()));
    }

    let mut score = 0;
    for (opponent, player) in moves {
        let round_score = play_round_part_1(&opponent, &player);
        score += round_score;
    }

    println!("Part 1: {}", score);

    let mut score = 0;
    for (opponent, player) in moves {
        let round_score = play_round_part_2(&opponent, &player);
        score += round_score;
    }

    println!("Part 2: {}", score);
}

fn play_round_part_1(opponent: &str, player: &str) -> i32 {
    let move_score_map = [
        (("Rock", "Scissors"), 0),
        (("Rock", "Paper"), 6),
        (("Rock", "Rock"), 3),
        (("Paper", "Rock"), 0),
        (("Paper", "Scissors"), 6),
        (("Paper", "Paper"), 3),
        (("Scissors", "Paper"), 0),
        (("Scissors", "Rock"), 6),
        (("Scissors", "Scissors"), 3),
    ];
    let opponent_move = move_map(opponent);
    let player_move = move_map(player);
    let score = move_score_map
        .iter()
        .find(|((a, b), _)| a == &opponent_move && b == &player_move)
        .unwrap()
        .1;
    score_map(&player_move) + score
}

fn play_round_part_2(opponent: &str, player: &str) -> i32 {
    let move_score_map = [
        (("Rock", "Scissors"), 6),
        (("Rock", "Paper"), 0),
        (("Rock", "Rock"), 3),
        (("Paper", "Rock"), 6),
        (("Paper", "Scissors"), 0),
        (("Paper", "Paper"), 3),
        (("Scissors", "Paper"), 6),
        (("Scissors", "Rock"), 0),
        (("Scissors", "Scissors"), 3),
    ];
    let opponent_move = move_map(opponent);
    let player_move = recommended_move_for_result_map(&opponent_move, &result_map(player));
    let score = move_score_map
        .iter()
        .find(|((a, b), _)| a == &player_move && b == &opponent_move)
        .unwrap()
        .1;
    score + score_map(&player_move)
}

fn move_map(move_str: &str) -> &str {
    match move_str {
        "Rock" => "Rock",
        "Paper" => "Paper",
        "Scissors" => "Scissors",
        _ => panic!("Invalid move"),
    }
}

fn result_map(move_str: &str) -> &str {
    match move_str {
        "Rock" => "Lose",
        "Paper" => "Win",
        "Scissors" => "Tie",
        _ => panic!("Invalid move"),
    }
}

fn recommended_move_for_result_map(opponent_move: &str, result: &str) -> &str {
    match (opponent_move, result) {
        ("Rock", "Win") => "Paper",
        ("Rock", "Lose") => "Scissors",
        ("Rock", "Tie") => "Rock",
        ("Paper", "Win") => "Scissors",
        ("Paper", "Lose") => "Rock",
        ("Paper", "Tie") => "Paper",
        ("Scissors", "Win") => "Rock",
        ("Scissors", "Lose") => "Paper",
        ("Scissors", "Tie") => "Scissors",
        _ => panic!("Invalid move"),
    }
}

fn score_map(move_str: &str) -> i32 {
    match move_str {
        "Rock" => 0,
        "Paper" => 6,
        "Scissors" => 3,
        _ => panic!("Invalid move"),
    }
}